Implement NOVA-Worldgen

ExE-Boss opened the pull request:
This PR contains my implementation of NOVA-Worldgen wrappers and WorldManager.
Depends on NOVA-Team/NOVA-Core#229, NOVA-Team/NOVA-Core#231 and NOVA-Team/NOVA-Core#232.

To do list:
- [ ] ~~Implement `WorldInfo`~~ (Leave that for another PR)

Completed:
- [x] `WorldgenManager` dependency injection

ExE-Boss committed:
Update to latest snapshot

TODO: Wrappers

ExE-Boss committed:
Begin work on Minecraft 1.11 wrappers

ExE-Boss committed:
Work on WorldManager + Add License to all files.

ExE-Boss committed:
Publish initial work on wrappers + changes to EnumSelector

EnumSelector will be removed if NOVA-Team/NOVA-Core#232 is merged.

ExE-Boss committed:
Fixed bugs in FWWorldGenerator + Switched to the EnumSelector in NOVA-Core

Depends on: NOVA-Team/NOVA-Core#232

ExE-Boss committed:
Rolled back rarity change, as higher means more common, not less common.

Victorious3 commented on minecraft/1.11/src/main/java/nova/worldgen/wrapper/mc/forge/v1_11/wrapper/world/forward/FWWorldGenerator.java:
```diff
@@ -0,0 +1,104 @@
+				//  DEEPERER = 20 (15-25)
+				//  REALLYDEEP = 10 (5-15)
+
+				if (ore.oreLayers.allows(OreHeight.SURFACE))
```
No. That's a loop.

ExE-Boss replied:
Fixed now

ExE-Boss committed:
Changed ore generation height into a loop + Height calculation fix

RX14 commented on minecraft/1.11/src/main/java/nova/worldgen/wrapper/mc/forge/v1_11/wrapper/world/forward/FWWorldGenerator.java:
```diff
@@ -74,24 +76,18 @@ public void generate(Random random, int chunkX, int chunkZ, World world, IChunkG
-					oreHeightMap.put(j++, OreHeight.REALLYDEEP);
+				Iterator<OreHeight> iterator = Arrays.stream(OreHeight.values()).filter(ore.oreLayers::allows).iterator();
+
+				for (OreHeight height = iterator.next(); iterator.hasNext();) {
```
`.forEach` on the stream?

ExE-Boss replied:
Fixed now.
The main reason I did it that way initially, is because I made use of the local variable `j`, which must be final or effectively final to be used in lambda expressions.

ExE-Boss committed:
Get rid of generic for loop and oreHeightMap for good

ExE-Boss committed:
Move wrapper main class to proper package

Reason: Consistency with already implemented wrappers

ExE-Boss committed:
Get rid of local variable j in FWWorldGenerator for good

ExE-Boss committed:
Created Generable

Reason: Worldgen is not just ores.

ExE-Boss committed:
Fix dependency injection + Fix FWWorldGenerator

Additional changes: OreGenerationRegistry was replaced with
Registry<Ore>, as it was redundant.

ExE-Boss committed:
Build script fixes

ExE-Boss committed:
Remove dependency on NOVA-Team/NOVA-Core#227 + Build script fixes

ExE-Boss committed:
Package fixes

ExE-Boss committed:
More fixes

ExE-Boss committed:
More fixes (1.11)

ExE-Boss commented:
Split into #2 and #3
